cpp/src/arrow/compute/kernels/vector_run_end_encode.cc

```diff
@@ -0,0 +1,654 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <utility>
+
+#include "arrow/compute/api_vector.h"
+#include "arrow/compute/kernels/common_internal.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/ree_util.h"
+
+namespace arrow {
+namespace compute {
+namespace internal {
+
+template <typename ArrowType, bool has_validity_buffer>
+struct ReadValueImpl {
```

Comment:
Is it possible to use one of the existing typed visitors instead of a handrolled loop?

Reply:
I think I can use `visit_data_inline.h` for some steps but not everything and the readability of the loops might become worse. I will try to do it.

Reply:
Just a suggestion - if it wouldn't improve it (and it doesn't seem like it could be a variant of one of the visitors) then don't worry about it

Reply:
Can you take a look at the last commit? The implementation is longer and relies on a lot of inlining to be efficient.

Reply:
I decided against this solution and added a commit that reverts it.